### ISSUE TITLE
revamp async support to better align with Rx-style libraries, possibly unify on Mutiny

### ISSUE THREAD

**Issue by Ladicek** [smallrye/smallrye-fault-tolerance#504]
All Rx-style libraries have future-like types (`Uni`, `Single`, `Mono`) that we currently support via SmallRye Reactive Converters. Internally, we use `CompletionStage`. This has one big drawback: all the Rx-style future types are _lazy_, but `CompletionStage` is _eager_.

Creating an eager future from a lazy future is simple (just subscribe before handing out the result), while it is not possible to create a lazy future from an eager future (it's already running!). To support lazy futures, we have essentially 2 (or 3, if we're brave) options:

1. Use `Supplier<CompletionStage>`.
2. Use one of the Rx-style libraries directly.
3. Create our own lazy future type.

To be honest, both 1 and 3 are bad ideas. Perhaps we should just unify on Mutiny.

**Comment by Ladicek** [comment]
Turns out that it's enough to move off of SmallRye Reactive Converters to our own conversion API, which would construct a lazy type _not_ from `CompletionStage`, but from `Supplier<CompletionStage>`. Everything else may stay as is.

This unfortunately doesn't allow supporting cancellation, but the core implementations of `CompletionStage` strategies doesn't allow supporting cancellation either. I think cancellation support can be deferred.

**Comment by edeandrea** [comment]
Hi @Ladicek is there an example somewhere of how I might use this? Specifically with [this example](https://github.com/quarkusio/quarkus-super-heroes/blob/main/rest-fights/src/main/java/io/quarkus/sample/superheroes/fight/client/HeroClient.java) where I have a method that returns `CompletionStage` and then I have to do `Uni.createFrom().completionStage(this::getRandomHero)`.

GIven this fix, should I not have to do that anymore?

**Comment by Ladicek** [comment]
Hi @edeandrea, I remember we came up with this weird combination of `Uni` and `CompletionStage` to work around a couple of issues. I don't recall what those issues were exactly, but I'm pretty sure at least some of them were fixed. That includes resubscription, which is what `Uni.retry()` does under the hood -- that should definitely be fixed.

**Comment by edeandrea** [comment]
Yeah I think there were several issues we were trying to work around. Maybe I'll try to un-do the custom stuff I was doing and see what issues still exist and what issues seem to be fixed.

**Comment by edeandrea** [comment]
Although I don't think Quarkus 2.7.3 uses fault-tolerance 5.3.0 yet (looks like its using 5.2.1), so maybe I have to wait on this a bit.

**Comment by Ladicek** [comment]
Ah correct. Quarkus `main` has 5.3.1, so you can try with a snapshot.
